### PR TITLE
Cassandra GDPR migration guide.

### PR DESCRIPTION
**Pull request by DenysGonchar**
to attach to the test env cassandra you can use the next command:
```
docker run -it --rm -e SSL_CERTFILE=/cacert.pem -v "${PWD}/tools/ssl/ca/cacert.pem:/cacert.pem:ro" --link mongooseim-cassandra:cassandra cassandra:3.9 sh -c 'exec cqlsh "$CASSANDRA_PORT_9042_TCP_ADDR" --ssl'
```

### CODE REVIEW

File: doc/migrations/3.3.0_3.3.0plus.md [hunk 1]
```diff
@@ -47,8 +47,51 @@ You'll need to iterate over the whole `mam_muc_message` table with the following
 
 ### Cassandra
 
-TODO
+#### Step 1
+
+Please execute following CQL statements on your MIM database:
```
Comment (goddammit) [inline doc/migrations/3.3.0_3.3.0plus.md]: Please execute the following CQL statements on your MIM database:

File: doc/migrations/3.3.0_3.3.0plus.md [hunk 2]
```diff
@@ -47,8 +47,51 @@ You'll need to iterate over the whole `mam_muc_message` table with the following
...
+Now you have a schema that is compatible with MIM 3.3.0++ but isn't GDPR-compliant yet because the new column has no meaningful data.
+
+Please pick your favourite scripting/programming language and populate the new column with the help of a [dedicated script](jid-from-mam-muc-script.md).
+You'll need to iterate over the whole `mam_muc_message` table with the following algorithm:
+
+1. extract the whole `mam_muc_message` table:
+```
+SELECT * FROM mam_muc_message;
+```
+1. To make data extraction faster MongooseIM stores 2 copies of the message in the table:
```
Comment (goddammit) [inline doc/migrations/3.3.0_3.3.0plus.md]:
```suggestion
1. To make data extraction faster, MongooseIM stores 2 copies of the message in the table:
```

File: doc/migrations/3.3.0_3.3.0plus.md [hunk 3]
```diff
@@ -47,8 +47,51 @@ You'll need to iterate over the whole `mam_muc_message` table with the following
...
+```
+cqlsh:mongooseim> SELECT * FROM mam_muc_message WHERE id = 399582233150625537 ALLOW FILTERING;
 
+ room_jid                      | with_nick | id                 | from_jid | message                        | nick_name
+-------------------------------+-----------+--------------------+----------+--------------------------------+-----------
+ room-ad1d999b9e@muc.localhost |           | 399582233150625537 |     null | 0x8350000001...998de2fa8426837 |       Sid
+ room-ad1d999b9e@muc.localhost |       Sid | 399582233150625537 |     null | 0x8350000001...998de2fa8426837 |       Sid
+```
+2. The copy with an empty `with_nick` column must be updated.
+4. Extract the **sender's JID** from the `message` column in the same way as described in RDBMS migration section. By default cassandra backend uses `eterm` format.
```
Comment (goddammit) [inline doc/migrations/3.3.0_3.3.0plus.md]:
```suggestion
4. Extract the **sender's JID** from the `message` column in the same way as described in the RDBMS migration section. By default cassandra backend uses the `eterm` format.
```

File: doc/migrations/3.3.0_3.3.0plus.md [hunk 4]
```diff
@@ -47,8 +47,51 @@ You'll need to iterate over the whole `mam_muc_message` table with the following
...
+3. Update the `from_jid` column with the value of extracted **sender's JID**:
```
Comment (goddammit) [inline doc/migrations/3.3.0_3.3.0plus.md]:
```suggestion
3. Update the `from_jid` column with the value of the extracted **sender's JID**:
```

**Comment by mongoose-im** [comment]
[6703.1](https://travis-ci.org/esl/MongooseIM/jobs/547788056) / Erlang 20.3 / small_tests / 5eb6c186c40839242bdf56af4e79cfc5ed024a3a
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2341/6703/small_tests.20.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2341/6703/small_tests.20.3/small/ct_run.mongooseim@localhost.2019-06-19_15.56.51/lib.mongooseim.logs/run.2019-06-19_15.56.52/suite.log.html)

File: doc/migrations/3.3.0_3.3.0plus.md [hunk 5]
```diff
@@ -47,8 +47,51 @@ You'll need to iterate over the whole `mam_muc_message` table with the following
...
+#### Step 2
+
+Now you have a schema that is compatible with MIM 3.3.0++ but isn't GDPR-compliant yet because the new column has no meaningful data.
+
+Please pick your favourite scripting/programming language and populate the new column with the help of a [dedicated script](jid-from-mam-muc-script.md).
+You'll need to iterate over the whole `mam_muc_message` table with the following algorithm:
+
+1. extract the whole `mam_muc_message` table:
+```
+SELECT * FROM mam_muc_message;
```
Comment (fenek) [inline doc/migrations/3.3.0_3.3.0plus.md]: Pretty much impossible for large DBs. Maybe it's better to recommend retrieving data by `id` in slices?

File: doc/migrations/3.3.0_3.3.0plus.md [hunk 6]
```diff
@@ -47,8 +47,51 @@ You'll need to iterate over the whole `mam_muc_message` table with the following
...
+cqlsh:mongooseim> UPDATE mam_muc_message SET from_jid = 'username@server' WHERE id = 399582233150625537  AND with_nick = '' AND room_jid = 'room-ad1d999b9e@muc.localhost';
+cqlsh:mongooseim> SELECT * FROM mam_muc_message WHERE id = 399582233150625537 ALLOW FILTERING;
+
+ room_jid                      | with_nick | id                 | from_jid        | message                        | nick_name
+-------------------------------+-----------+--------------------+-----------------+--------------------------------+-----------
+ room-ad1d999b9e@muc.localhost |           | 399582233150625537 | username@server | 0x8350000001...998de2fa8426837 |       Sid
+ room-ad1d999b9e@muc.localhost |       Sid | 399582233150625537 |            null | 0x8350000001...998de2fa8426837 |       Sid
+
+```
+3.
```
Comment (fenek) [inline doc/migrations/3.3.0_3.3.0plus.md]: Leftover?

Comment (DenysGonchar) [reply doc/migrations/3.3.0_3.3.0plus.md]: nope, added intentionally to show which record is updated.

File: doc/migrations/3.3.0_3.3.0plus.md (hunk 5 above)
Comment (DenysGonchar) [reply doc/migrations/3.3.0_3.3.0plus.md]: yes, you right.

Comment (DenysGonchar) [reply doc/migrations/3.3.0_3.3.0plus.md]: well,  to filter by `id` we have to specify the key (some `room_jid` and `with_nick = ''`) or specify `ALLOW FILTERING`:
```
SELECT * FROM mongooseim.mam_muc_message LIMIT 10;
SELECT * FROM mongooseim.mam_muc_message where id > 399582468742569729 LIMIT 10  ALLOW FILTERING;
```

i believe `ALLOW FILTERING` is not the best idea, nor the room by room update.

cassandra drivers allow paging, i think this one should be used instead. but it's very driver specific to describe it here and CQL command in this case is still the same.

File: doc/migrations/3.3.0_3.3.0plus.md (hunk 6 above)
Comment (DenysGonchar) [reply doc/migrations/3.3.0_3.3.0plus.md]: ooops, i haven't noticed that `3.` :)

**Pull request merged**